File: code/jvm/host/src/main/java/pt/isel/host/config/JacksonConfig.java
package pt.isel.host.config;

import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import org.springframework.boot.autoconfigure.jackson.Jackson2ObjectMapperBuilderCustomizer;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import pt.isel.domain.messages.UpdatedMessage;
import pt.isel.domain.messages.UpdatedMessage.KeepAlive;
import pt.isel.domain.messages.UpdatedMessage.NewMessage;

@Configuration
public class JacksonConfig {

    @Bean
    public Jackson2ObjectMapperBuilderCustomizer jacksonCustomizer() {
        return builder -> {
            builder.mixIn(UpdatedMessage.class, UpdatedMessageMixin.class);
        };
    }

    @JsonTypeInfo(
            use = JsonTypeInfo.Id.NAME,
            include = JsonTypeInfo.As.PROPERTY,
            property = "type"
    )
    @JsonSubTypes({
            @JsonSubTypes.Type(value = NewMessage.class, name = "new-message"),
            @JsonSubTypes.Type(value = KeepAlive.class, name = "keep-alive")
    })
    public abstract static class UpdatedMessageMixin {
    }
}

File: code/jvm/host/src/main/kotlin/pt/isel/host/config/JacksonConfig.kt
package pt.isel.host.config

import com.fasterxml.jackson.annotation.JsonSubTypes
import com.fasterxml.jackson.annotation.JsonTypeInfo
import com.fasterxml.jackson.module.kotlin.KotlinModule
import org.springframework.boot.autoconfigure.jackson.Jackson2ObjectMapperBuilderCustomizer
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import pt.isel.domain.messages.KeepAlive
import pt.isel.domain.messages.NewMessage
import pt.isel.domain.messages.UpdatedMessage

@Configuration
class JacksonConfig {
    @Bean
    fun jacksonCustomizer(): Jackson2ObjectMapperBuilderCustomizer =
        Jackson2ObjectMapperBuilderCustomizer { builder ->
            builder.modulesToInstall(KotlinModule.Builder().build())
            builder.mixIn(UpdatedMessage::class.java, UpdatedMessageMixin::class.java)
        }

    @JsonTypeInfo(
        use = JsonTypeInfo.Id.NAME,
        include = JsonTypeInfo.As.PROPERTY,
        property = "type",
    )
    @JsonSubTypes(
        JsonSubTypes.Type(value = NewMessage::class, name = "new-message"),
        JsonSubTypes.Type(value = KeepAlive::class, name = "keep-alive"),
    )
    abstract class UpdatedMessageMixin
}
